### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Phillip-SICE**
First pull request

### CODE REVIEW

File: UDPClientGUI/MainWindow.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace UDPCommGUI
+{
+    /// <summary>
+    /// Interaction logic for MainWindow.xaml
+    /// </summary>
+    public partial class MainWindow : Window
+    {
+
+        MyUDPClient client = new MyUDPClient();
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: Use `var` where possible. In C#, if you have the type on the right hand side you usually don't need the type on the left hand side.

There are a few cases where you'd explicitly define the type, here are 2, but there may be more.
- When it's not initialised at the time of declaration
- For `Actions/Func` - check these out

File: UDPClientGUI/MainWindow.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        private void IPChanged(object sender, TextChangedEventArgs e)
+        {
+            client.ConnectionIP = IPbox.Text;
+        }
+
+        private void PortChanged(object sender, TextChangedEventArgs e)
+        {
+            try
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: Why is there a try/finally? You could just use int.TryParse

Note that `Int32` is the same as `int` but you'd rarely see the first one.

Have a read of `ref`, `in` and `out` while you're at it

File: UDPClientGUI/MyUDPClient.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        private string connectionIP;
+        private int connectionPort;
+        private string inputMessage;
+
+
+        public string ConnectionIP
+        {
+            get { return connectionIP; }
+            set { connectionIP = value; }
+        }
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: With Autoproperties, if you're not doing anything at the time of getting/setting, you don't need to have it this way.

`public string ConnectionIP { get; set; }`

Where you would use a backing field (like it's done here), is if it's a different data type to the Property or if you have extra logic around getting/setting.

Example:
```
private int myVariable;

public string MyVariable => myVariable.ToString();
```
While you're at it, check out `Expression Bodied Members`

File: UDPClientGUI/MyUDPClient.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        public int ConnectionPort
+        {
+            get { return connectionPort; }
+            set { connectionPort = value; }
+        }
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: Please fix all instances as above.

File: UDPClientGUI/MyUDPClient.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        public void Connect()
+        {
+            this.Connect(ConnectionIP, ConnectionPort);
+        }
+
+        public void DisplayInfo()
+        {
+            //
+        }
+
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: If it's not being used, it should be cleaned up.

Good idea to check https://www.martinfowler.com/bliki/Yagni.html

File: UDPServer/IServerInterface.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace UDPServer
+{
+    interface IServerInterface
+    {
+        string ConnectionIP
+        {
+            get;
+            set;
+        }
```
Comment (sk-sice) [inline UDPServer/IServerInterface.cs]: The correct formatting is `string ConnectionIP { get; set; }`

File: UDPServer/MyUDPServer.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            get { return connectionPort; }
+            set { connectionPort = value; }
+        }
+
+        public string ReceivedData
+        {
+            get { return receivedData; }
+            set { receivedData = value; }
+        }
```
Comment (sk-sice) [inline UDPServer/MyUDPServer.cs]: As above.

File: UDPServer/MyUDPServer.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            client = new UdpClient(ConnectionPort);
+        }
+
+        public void GetData()
+        {
+            // IPEndPoint remoteIpEndPoint = new IPEndPoint(IPtoLong(ConnectionIP), ConnectionPort);
+            IPEndPoint remoteIpEndPoint = new IPEndPoint(IPAddress.Any, ConnectionPort);
+            byte[] data = client.Receive(ref remoteIpEndPoint);
+            ReceivedData = Encoding.ASCII.GetString(data);
+        }
```
Comment (sk-sice) [inline UDPServer/MyUDPServer.cs]: use var

File: UDPServer/MyUDPServer.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            string[] IPs = IP.Split('.');
+            long result = 0;
+            int i = 24;
+            foreach(string s in IPs)
+            {
+                result += long.Parse(s) * 2 ^ i;
+                i -= 8;
+            }
+
+            return result;
```
Comment (sk-sice) [inline UDPServer/MyUDPServer.cs]: Why is this required?

**Review by sk-sice — CHANGES_REQUESTED** [review]
Please fix the above

Also please move the IP Address and Port to the configuration file.

File: UDPClientGUI/MainWindow.xaml.cs (hunk 1 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: I am getting error message "The contextual keyword 'var' may only appear within a local variable declaration or in script code" as follows:
image

File: UDPClientGUI/MainWindow.xaml.cs (hunk 2 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: Changed as suggested.

File: UDPClientGUI/MyUDPClient.cs (hunk 3 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Changed as suggested.

File: UDPClientGUI/MyUDPClient.cs (hunk 4 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Changed as suggested.

File: UDPClientGUI/MyUDPClient.cs (hunk 5 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Removed from interface and class.

File: UDPServer/IServerInterface.cs (hunk 6 above)
Comment (Phillip-SICE) [reply UDPServer/IServerInterface.cs]: Changed for both interface and class.

File: UDPServer/MyUDPServer.cs (hunk 7 above)
Comment (Phillip-SICE) [reply UDPServer/MyUDPServer.cs]: Changed for both interface and class

File: UDPServer/MyUDPServer.cs (hunk 8 above)
Comment (Phillip-SICE) [reply UDPServer/MyUDPServer.cs]: Changed as commented.

File: UDPServer/MyUDPServer.cs (hunk 9 above)
Comment (Phillip-SICE) [reply UDPServer/MyUDPServer.cs]: It was used for creating IPEndPoint. Method has been removed and IPAddress.Parse() is used instead

File: UDPClientGUI/MainWindow.xaml.cs (hunk 1 above)
Comment (sk-sice) [reply UDPClientGUI/MainWindow.xaml.cs]: Ignore that, my mistake - can't use `var` with class fields. You can resolve this conversation.

File: UDPClientGUI/MainWindow.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        private void IPChanged(object sender, TextChangedEventArgs e)
+        {
+            client.ConnectionIP = IPbox.Text;
+        }
+
+        private void PortChanged(object sender, TextChangedEventArgs e)
+        {
+            int parseOut;
+            if (int.TryParse(PortBox.Text, out parseOut))
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: Ideally if you couldn't parse it you'd have a Dialog box popup stating that it failed.
But this can be done later.

File: UDPServer/MainWindow.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            server.ConnectionPort = Int32.Parse(PortBox.Text);
+        }
+
+        private void ConnectButtonClicked(object sender, RoutedEventArgs e)
+        {
+            if(!server.ConnectionStatus)
+            {
+                server.Connect();
+                listeningThread = new Thread(new ThreadStart(GetData));
```
Comment (sk-sice) [inline UDPServer/MainWindow.xaml.cs]: Replace Thread with `async/await`.

Use an async void ConnectButtonClicked to execute GetData().

Get Data() should also be an `async Task` method **without** the `while(true)`

File: UDPServer/MyUDPServer.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+        public bool ConnectionStatus { get; set; }
+
+        public ObservableCollection<string> MessageReceived { get; set; }
+
+        public int MessageReceivedNumber { get; set; }
+
+        public int DataAvailable
+        {
+            get { return client.Available; }
```
Comment (sk-sice) [inline UDPServer/MyUDPServer.cs]: Use expression bodied getter

File: UDPClientGUI/MainWindow.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace UDPCommGUI
+{
+    /// <summary>
+    /// Interaction logic for MainWindow.xaml
+    /// </summary>
+    public partial class MainWindow : Window
+    {
+
+        MyUDPClient client = new MyUDPClient();
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: `private readonly`

File: UDPClientGUI/MainWindow.xaml.cs [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+            client.SendMessage();
+        }
+
+        private void DisConnectClicked(object sender, RoutedEventArgs e)
+        {
+            client.Disconnect();
+            ChangeStatus();
+        }
+
+        private void ChangeStatus()
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: Avoid multiple levels of nesting in a method unless absolutely necessary

You could check the connection status first and set the Text and Foreground first.

Then have a Dispatcher.Invoke

File: UDPClientGUI/MyUDPClient.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                client.Close();
+                this.ConnectionStatus = false;
+            }
+        }
+
+        public void SendMessage()
+        {
+            if(ConnectionStatus)
+            {
+                byte[] message = Encoding.ASCII.GetBytes(InputMessage);
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: Use `var`

File: UDPClientGUI/MyUDPClient.cs [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+            this.client = new UdpClient();
+            client.Connect(ConnectionIP, ConnectionPort);
+            this.ConnectionStatus = true;
+        }
+
+        public void Disconnect()
+        {
+            if(ConnectionStatus)
+            {
+                client.Close();
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: It's a good idea to Dispose of the UDP client when the application closes if you haven't already.

File: UDPClientGUI/MyUDPClient.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace UDPCommGUI
+{
+    class MyUDPClient : IClientInterface
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: Also implement IDisposable so you can dispose of it later

File: UDPServer/IServerInterface.cs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+        int ConnectionPort { get; set; }
+
+        string ReceivedData { get; set; }
+
+        void Connect();
+
+        void Disconnect();
+
+        //Task GetData();
```
Comment (sk-sice) [inline UDPServer/IServerInterface.cs]: If this is not used, remove it. Always make sure that commented out code is removed before committing.

File: UDPServer/MainWindow.xaml.cs (hunk 11 above)
Comment (sk-sice) [reply UDPServer/MainWindow.xaml.cs]: You may need a Task.Run with a cancellation token when you want to stop listening.

File: UDPClientGUI/MainWindow.xaml.cs [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            client.InputMessage = MessageBox.Text;
+        }
+
+        private void SendButtonClicked(object sender, RoutedEventArgs e)
+        {
+            client.SendMessage();
+        }
+
+        private void DisConnectClicked(object sender, RoutedEventArgs e)
```
Comment (sk-sice) [inline UDPClientGUI/MainWindow.xaml.cs]: Disconnect instead of DisConnect

File: UDPServer/MyUDPServer.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        public async Task GetData()
+        {
+            var remoteIpEndPoint = new IPEndPoint(IPAddress.Parse(ConnectionIP), ConnectionPort);
+            //byte[] data = client.Receive(ref remoteIpEndPoint);
+            var data = await client.ReceiveAsync();
+            ReceivedData = Encoding.ASCII.GetString(data.Buffer);
+            this.MessageReceivedNumber++;
+            this.MessageReceived.Add(this.MessageReceivedNumber.ToString() + ". " + ReceivedData);
```
Comment (sk-sice) [inline UDPServer/MyUDPServer.cs]: Use `string` interpolation

**Review by sk-sice — APPROVED** [review]

**Review by sk-sice — CHANGES_REQUESTED** [review]
Please fix the few things mentioned above. 

The main ones are to use a Task.Run with a Cancellation token instead of managing a thread yourself and to dipose of the client on shutdown ensuring this is done through an IDisposable interface

File: UDPClientGUI/MyUDPClient.cs (hunk 16 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Implemented for both server and client

File: UDPClientGUI/MyUDPClient.cs (hunk 17 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Implemented for both server and client

File: UDPServer/IServerInterface.cs (hunk 18 above)
Comment (Phillip-SICE) [reply UDPServer/IServerInterface.cs]: Removed.

File: UDPClientGUI/MainWindow.xaml.cs (hunk 19 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: Fixed

File: UDPClientGUI/MainWindow.xaml.cs (hunk 10 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: Implemented MessageBox popup

File: UDPServer/MainWindow.xaml.cs (hunk 11 above)
Comment (Phillip-SICE) [reply UDPServer/MainWindow.xaml.cs]: Implemented

File: UDPClientGUI/MainWindow.xaml.cs (hunk 13 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: Changed as commented

File: UDPClientGUI/MainWindow.xaml.cs (hunk 14 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MainWindow.xaml.cs]: Fixed as commented.

File: UDPClientGUI/MyUDPClient.cs (hunk 15 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Fixed as commented

File: UDPServer/MyUDPServer.cs (hunk 20 above)
Comment (Phillip-SICE) [reply UDPServer/MyUDPServer.cs]: Changed as commented

**Review by sk-sice — APPROVED** [review]

File: UDPServer/MainWindow.xaml.cs [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        bool ipValid;
+        CancellationTokenSource ctSource;
+        CancellationToken token;
+
+        public MainWindow()
+        {
+            portValid = false;
+            ipValid = false;
+            InitializeComponent();
+            //ctSource = new CancellationTokenSource();
```
Comment (sk-sice) [inline UDPServer/MainWindow.xaml.cs]: Please don't leave commented code behind. It will confuse yourself and people who code after you.

File: UDPClientGUI/IClientInterface.cs [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+        string InputMessage { get; set; }
+
+        bool ConnectionStatus { get; }
+
+        void Connect();
+
+        void Disconnect();
+
+        void SendMessage();
+
```
Comment (sk-sice) [inline UDPClientGUI/IClientInterface.cs]: Minor thing, watch the spacing (we'll sort this out using Style rules later on).

File: UDPClientGUI/MyUDPClient.cs [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace UDPCommGUI
+{
+    class MyUDPClient : IClientInterface
```
Comment (sk-sice) [inline UDPClientGUI/MyUDPClient.cs]: MyUDPClient is better named as "SiceUDPClient" - not a major thing.

**Review by sk-sice — CHANGES_REQUESTED** [review]
Please remove commented code

File: UDPServer/MainWindow.xaml.cs (hunk 21 above)
Comment (Phillip-SICE) [reply UDPServer/MainWindow.xaml.cs]: Removed.

File: UDPClientGUI/IClientInterface.cs (hunk 22 above)
Comment (Phillip-SICE) [reply UDPClientGUI/IClientInterface.cs]: Fixed.

File: UDPClientGUI/MyUDPClient.cs (hunk 23 above)
Comment (Phillip-SICE) [reply UDPClientGUI/MyUDPClient.cs]: Changed as commented.
